lib/breadcrumbs.ts

```diff
@@ -0,0 +1,182 @@
+import type { AsyncLocalStorage } from "async_hooks";
+import type { Breadcrumb, InternalBreadcrumb } from "./types";
+import type { Request, Response } from "express";
+const debug = require("debug")("raygun").bind(null, "[breadcrumbs]");
+
+let asyncLocalStorage: AsyncLocalStorage<InternalBreadcrumb[]> | null = null;
+
+try {
+  asyncLocalStorage = new (require("async_hooks").AsyncLocalStorage)();
+  debug("initialized successfully");
+} catch (e) {
+  debug(
+    "failed to load async_hooks.AsyncLocalStorage - initialization failed\n",
+    e,
+  );
+}
+
+type SourceFile = {
+  fileName: string;
+  functionName: string;
+  lineNumber: number | null;
+};
+
+function returnCallerSite(
+  err: Error,
+  callsites: NodeJS.CallSite[],
+): SourceFile | null {
+  for (const callsite of callsites) {
+    const fileName = callsite.getFileName() || "";
+
+    if (fileName.startsWith(__dirname)) {
+      continue;
+    }
+
+    return {
+      fileName: fileName,
+      functionName: callsite.getFunctionName() || "<anonymous>",
+      lineNumber: callsite.getLineNumber(),
+    };
+  }
+
+  return null;
+}
+
+function getCallsite(): SourceFile | null {
+  const originalPrepareStacktrace = Error.prepareStackTrace;
+
+  Error.prepareStackTrace = returnCallerSite;
+
+  // Ignore use of any, required for captureStackTrace
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const output: any = {};
+
+  Error.captureStackTrace(output);
+
+  const callsite = output.stack;
+  Error.prepareStackTrace = originalPrepareStacktrace;
+  return callsite;
+}
+
+export function addBreadcrumb(
+  breadcrumb: string | Breadcrumb,
+  type: InternalBreadcrumb["type"] = "manual",
+) {
+  const crumbs = getBreadcrumbs();
+
+  debug("using breadcrum store:", crumbs);
+
+  if (!crumbs) {
+    return;
+  }
+
+  if (typeof breadcrumb === "string") {
+    const expandedBreadcrumb: Breadcrumb = {
+      message: breadcrumb,
+      level: "info",
+      category: "",
+    };
+
+    breadcrumb = expandedBreadcrumb;
+  }
+
+  const callsite = getCallsite();
+
+  const internalCrumb: InternalBreadcrumb = {
+    ...(breadcrumb as Breadcrumb),
+    category: breadcrumb.category || "",
+    message: breadcrumb.message || "",
+    level: breadcrumb.level || "info",
+    timestamp: Number(new Date()),
+    type,
+    className: callsite?.fileName,
+    methodName: callsite?.functionName,
+
+    // TODO - do we need to do any source mapping?
+    lineNumber: callsite?.lineNumber || undefined,
+  };
+
+  debug("[breadcrumbs.ts] recorded breadcrumb:", internalCrumb);
+
+  crumbs.push(internalCrumb);
+}
+
+export function addRequestBreadcrumb(request: Request, response: Response) {
```

Comment:
logs an ExpressJS request as a Breadcrumb, used by the `expressHandlerBreadcrumbs`

Reply:
Should this express-specific behaviour be in a generic `breadcrumbs` module then or maybe in a separate one for handling Express?

Reply:
Or should at least the function name be less generic then?

Reply:
Looking a bit into it, I think that `Request` is an object from the standard JS library and not an Express object per se, although this method is only used by the `expressHandlerBreadcrumbs` method, it could be used to register non-ExpressJS requests.

We could probably expose an `addRequestBreadcrumb` as well in the `raygun.js` to be used in other web frameworks, if there is a use for it.

Reply:
Moved to `raygun.breadcrumbs.express.ts`